src/Select/MenuWrapper.js

```diff
@@ -60,21 +61,28 @@ class MenuWrapper extends Component {
         `
 
         return ReactDOM.createPortal(
-            <div className={className} ref={menuRef}>
-                <Card className={cardClassName}>{children}</Card>
-
-                {styles}
-
-                <style jsx>{`
-                    div {
-                        position: absolute;
-                        z-index: ${layers.applicationTop};
-                        top: ${top};
-                        left: ${left};
-                        width: ${width};
-                    }
-                `}</style>
-            </div>,
+            <ZIndexConsumer
+                propZIndex={zIndex}
+                defaultZIndex={layers.applicationTop}
+            >
```

Comment:
I am slightly confused by this usage. I've never seen `props` passed to the consumer this way.

Usually I would expect something like:

```
class ZIndexProvider extends React.Component {
  state = {
    zIndex: 0,
    setIndex = zIndex => this.setState({ zIndex })
  }

  render() {
    <ZIndexContext.Provider value={this.state}>
        {children}
    </ZIndexContext.Provider>
  }
}
```

```
<ZIndexConsumer>
  {({computedZIndex, setZIndex}) => {
     if (props.zIndex !== computedZIndex) {
        setZIndex(props.zIndex)
      }

    return (
      <div>...</div>
    )
  })
</ZIndexConsumer>
```

Reply:
Even though I will be transitioning to hooks now, I would still like to explain how I ended up with this, perhaps unconventional, but very practical approach:
- In essence, all that the context needs to provide is the actual z-index that a component needs to build upon, so it made sense to let the context value be a simple number.
- The calculation of the z-index value for the "next layer" is based on the current z-index context, but does not actually affect the value of that context. As such it made more sense to extract that logic into the consumer rather than the provider.

I did study the docs and saw examples such as the one you provided, but I think the implementation I chose in the end is a better conceptual fit with the problem we are trying to solve. But yeah, I just made it up as I went along, I'm not following a pattern of any sorts...

Reply:
Perhaps to make what I was trying to explain a bit more tangible. In your example, there's this line:
```
setIndex = zIndex => this.setState({ zIndex }
```
But actually we don't want to mutate the z-index from that layer. We want only want built upon that. The z-index of that layer should be constant. This way we can use this in a hierarchical tree-structure.